### DISCUSSION TITLE
Stats/Spotify Wrapped style page?

### DISCUSSION THREAD

**Discussion by yllekz** [navidrome/navidrome#3524] [Ideas]
Would anyone be interested in a feature where Navidrome could generate a "Spotify Wrapped" style of song statistics for sharing purposes? Does such a feature already exist?

**Comment by kgarner7** [comment]
This isn't _really_ possible with Navidrome (not easily anyway), as the database just stores total play counts, not individual listens (so you can't know when a track was listened). You could of course write an application that collects these stats each year and then work with that, but I doubt it'll come in Navidrome itself.

Alternatively, if you scrobble to last.fm or listenbrainz (or self-hosted Maloja), you can probably accomplish the same thing (but better). Last.fm had a wrapped for last year (albeit later than other platforms), so you can wait for that.

**Comment by metalheim** [comment]
To add to kgarner7's answer: If you scrobble to [listenbrainz](https://listenbrainz.org/) (a libre alternative to last.fm), you also get a "Year in Music" review which is similar to spotify's "wrapped".
It usually drops on 05. January the following year.

But the beauty of either last.fm / listenbrainz / maloja is that you can see your listening statistics at any time (no need to wait for spotify to "drop" a "wrapped") for any time period you like.

**Comment by Menghini** [comment]
Question: I have done my own "Navidrome Wrapped" on my server by manually querying the database with this query:
```
SELECT annotation.item_id, media_file.title, media_file.album, annotation.play_count, media_file.duration
FROM annotation
JOIN media_file ON annotation.item_id = media_file.id
WHERE annotation.item_type = 'media_file' AND user_id = 'YourUserID' AND annotation.play_count != 0
ORDER BY annotation.play_count DESC, media_file.title ASC;
```

And I take the difference between each `annotation.item_id` between last year and this year's `annotation.play_count` Wouldn't this be enough? Or is this not accurate?

Of course, in order to do what OP has asked, you would need to keep track of the amount of listens for that given year, which would require a new column.

**Reply by kgarner7** [reply]
This works if 
1. You keep a diff year after year
2. You only want a snapshot over the entire year (and say, don't want to know which day was the most)

If you just want the headline numbers though, this is a nice and reasonably accurate calculation (until the BFR changes the database schema, then you'll need to update your queries again a bit).

**Reply by Menghini** [reply]
I agree. I would be happy to implement this, but it obvsiouly wouldn't work for this year since it would have to start counting Jan 1.

The two ways I am thinking of: 1) have a column in `annotation` for 2025 plays (but then you would need a 2026 plays, 2027 plays, etc) or 2) have a column in `annotation` for this year's plays and reset it when it becomes the new year. I could care less on what was played on a particular day. It just depends on how much info you want, I personally only cared about the number of times I played a song and wanted my top 10. Spotify is more in depth with favorite artists and this year, implemented it as a per month, but it usally has been per year. When I see friend's share their Spotify Wrapped they only share their per year page. Thoughts?

Here is an example of what it could look like. This is mine for 2024 I shared on my social media (along with Spotify like everyone else), but I manually ran the query and knew the amount of times I played the songs last year (I ran the query last year and this year and saved it). I just made this in Google Sheets. Would be cleaned up and integrated in Navidrome. 
![Navidrome Wrapped 2024](https://github.com/user-attachments/assets/f5be469b-4a92-4277-9d15-57363e760bd6)

In theory, you could also keep a backup of the database around midnight on Jan 1 and compare the database with that one using a similar query.
